### PR TITLE
[FLINK-21087][runtime][checkpoint] StreamTask waits for all the pending checkpoints to finish before finished

### PR DESCRIPTION
**Pull request by gaoyunhaii**
## What is the purpose of the change

This PR makes `StreamTask` to wait for all the pending checkpoints to finish before finished. Previously checkpoints after tasks finished is not supported, and aborting all the pending checkpoints when finished is reasonable. However, after this feature is considering, it would be normal to have tasks finished when taking checkpoints, and it should be more reasonable to wait for the pending checkpoints to finish.

Since for each checkpoint, the synchronous part is done immediately when triggered, thus we need only wait for all the pending `AsyncCheckpointRunnable` to finish.

## Brief change log

- 56e89104538519e2a74f3276149da0b1372bd024 implements the waiting for StreamTask.

## Verifying this change

This change added tests and can be verified via added UT.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): **no**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: **no**
  - The serializers: **no**
  - The runtime per-record code paths (performance sensitive):  **no**
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: **no**
  - The S3 file system connector: **no**

## Documentation

  - Does this pull request introduce a new feature? **no**
  - If yes, how is the feature documented? **not applicable**

### CODE REVIEW

**Comment by gaoyunhaii** [comment]
Hi @dawidwys , could you also have a look at this PR when convenient~?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/AsyncCheckpointRunnable.java [hunk 1]
```diff
@@ -110,8 +109,6 @@ public void run() {
         FileSystemSafetyNet.initializeSafetyNetForThread();
         try {
 
-            registerConsumer.accept(this);
```
Comment (dawidwys) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/AsyncCheckpointRunnable.java:113]: Is this change necessary? We are changing the time when the runnable is registered slightly. Before the change it was registered when run, now when the runnable is created.

Comment (gaoyunhaii) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/AsyncCheckpointRunnable.java:113]: I think it might be necessary since otherwise there might be the case

1. We trigger a checkpoint and schedule its asynchronous part.
2. We call `waitForPendingCheckpoints()`. However, the asynchronous part is not executed yet.
3. After the call is done, the asynchronous part is executed and register the runnable. 

Then this runnable would be missed~

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java [hunk 2]
```diff
@@ -1864,6 +1866,61 @@ static void processMailTillCheckpointSucceeds(
         testHarness.getTaskStateManager().getWaitForReportLatch().await();
     }
 
+    @Test
+    public void testWaitingForPendingCheckpointsOnFinished() throws Exception {
```
Comment (dawidwys) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java:1870]: Currently, the test passes both with and without waiting for the pending checkpoints.

Moreover I think there is a lot of timing assumptions that need to be met for the test to verify the intended behaviour. I could not think of a good way to fix all of them (would be best to remove the remaining `Thread#sleep()`), but what do you think about such test:

```
    @Test
    public void testWaitingForPendingCheckpointsOnFinished() throws Exception {
        OneShotLatch asyncCheckpointExecuted = new OneShotLatch();
        OneShotLatch canCheckpointBeAcknowledged = new OneShotLatch();
        OneShotLatch invokeCompleted = new OneShotLatch();
        TestCheckpointResponder responder =
                new TestCheckpointResponder() {
                    @Override
                    public void acknowledgeCheckpoint(
                            JobID jobID,
                            ExecutionAttemptID executionAttemptID,
                            long checkpointId,
                            CheckpointMetrics checkpointMetrics,
                            TaskStateSnapshot subtaskState) {

                        try {
                            asyncCheckpointExecuted.trigger();
                            canCheckpointBeAcknowledged.await();
                        } catch (Exception e) {
                            throw new RuntimeException(e);
                        }
                    }
                };

        CompletableFuture<Void> taskClosed =
                CompletableFuture.runAsync(
                        () -> {
                            try (StreamTaskMailboxTestHarness<String> harness =
                                    new StreamTaskMailboxTestHarnessBuilder<>(
                                                    OneInputStreamTask::new,
                                                    BasicTypeInfo.STRING_TYPE_INFO)
                                            .addInput(BasicTypeInfo.STRING_TYPE_INFO, 3)
                                            .modifyStreamConfig(
                                                    config -> config.setCheckpointingEnabled(true))
                                            .setCheckpointResponder(responder)
                                            .setupOperatorChain(new EmptyOperator())
                                            .finishForSingletonOperatorChain(
                                                    StringSerializer.INSTANCE)
                                            .build()) {

                                harness.streamTask
                                        .getCheckpointCoordinator()
                                        .setEnableCheckpointAfterTasksFinished(true);

                                harness.streamTask.triggerCheckpointOnBarrier(
                                        new CheckpointMetaData(1, 101),
                                        CheckpointOptions.forCheckpointWithDefaultLocation(),
                                        new CheckpointMetricsBuilder()
                                                .setBytesProcessedDuringAlignment(0L)
                                                .setAlignmentDurationNanos(0L));

                                harness.waitForTaskCompletion();
                                invokeCompleted.trigger();
                                harness.finishProcessing();
                            } catch (Exception e) {
                                e.printStackTrace();
                            }
                        },
                        Executors.newSingleThreadExecutor());

        asyncCheckpointExecuted.await();
        invokeCompleted.await();
        Thread.sleep(500); // give some potential time for the task to finish before the
        // checkpoint is acknowledged
        assertFalse(taskClosed.isDone());
        canCheckpointBeAcknowledged.trigger();
        taskClosed.get();
    }
```

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java [hunk 3]
```diff
@@ -406,6 +406,31 @@ public void initInputsCheckpoint(long id, CheckpointOptions checkpointOptions)
...
+            asyncCheckpointRunnables = new ArrayList<>(checkpoints.values());
+        }
+
+        // Waits for each checkpoint independently.
+        asyncCheckpointRunnables.forEach(
+                ar -> {
+                    try {
+                        ar.getFinishedFuture().get();
+                    } catch (Exception e) {
+                        LOG.error(
```
[the remark this answers is not in the document]
Comment (gaoyunhaii) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:425]: I think here perhaps here we do not want to cause failover if one async runnable completed exceptionally ? For normal execution the async runnable fail seems also only fail the checkpoint but do not cause failover~

[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:425]: I get the original intention, but to implement that we need changes in the `AsyncCheckpointRunnable`, don't we? Currently the `getFinishedFuture` never finishes exceptionally. We only ever call `finishedFuture.complete(null)` in the `AsyncCheckpointRunnable`. The only exception can appear from interrupting the thread that is calling the `waitForPendingCheckpoints`. Or am I wrong?

[the remark this answers is not in the document]
Comment (gaoyunhaii) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:425]: No, that should be indeed a mistake, I'll modify this part~

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/AsyncCheckpointRunnable.java (hunk 1 above)
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/AsyncCheckpointRunnable.java:113]: BTW, sorry I missed one comment in the previous review (not sure how). Let's inline the regsiterConsumer() and remove the method. It is unnecessary to go through the `Consumer` here.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java [hunk 4]
```diff
@@ -527,20 +552,22 @@ private void finishAndReportAsync(
...
                         metrics,
                         System.nanoTime(),
                         taskName,
-                        registerConsumer(),
                         unregisterConsumer(),
                         env,
                         asyncExceptionHandler,
-                        isRunning));
+                        isRunning);
+        registerConsumer().accept(asyncCheckpointRunnable);
```
Comment (dawidwys) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:566]: BTW, sorry I missed one comment in the previous review (not sure how). Let's inline the regsiterConsumer() and remove the method. It is unnecessary to go through the Consumer here.

I believe we can replace it with:
```
registerAsyncCheckpointRunnable(
                        asyncCheckpointRunnable.getCheckpointId(), asyncCheckpointRunnable);
```

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java (hunk 2 above)
Comment (gaoyunhaii) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java:1869]: Sorry for not check if the test is work in reverse cases... It seems indeed not easy to accurately test "wait", and I think the suggested implementation is very good. I have changed  the test to this implementation~

**Comment by gaoyunhaii** [comment]
Very thanks @dawidwys for the review! I have updated the PR according to the comments~

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java [hunk 5]
```diff
@@ -563,24 +563,19 @@ private void finishAndReportAsync(
                         env,
                         asyncExceptionHandler,
                         isRunning);
-        registerConsumer().accept(asyncCheckpointRunnable);
+
+        try {
```
Comment (dawidwys) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:567]: The `try-catch` is unnecessary here. You can safely add the IOException to the method signature. I think the only reason for wrapping it before was to match the `Consumer` interface.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:425]: Now I am a bit afraid we log the exception a few times too many. It is logged on an info level in the `CheckpointAsyncRunnable` then again in the `StreamTask#handleAsyncException` (somewhere deeper in the stack).

WDYT about dropping the log? Or how about we log it on the `debug` level`?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java (hunk 5 above)
Comment (gaoyunhaii) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java]: Ah, sorry for ignore this, I have removed the exception wrapper~

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (gaoyunhaii) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java]: The log is indeed repeat, Perhaps we first change it to the `debug` level ?

**Comment by dawidwys** [comment]
@flinkbot run azure

**Review by dawidwys — APPROVED** [review]
LGTM, merging
